src/main/java/com/google/api/gax/grpc/RetryingCallable.java

```diff
@@ -29,226 +29,151 @@
+import com.google.api.gax.retrying.TimedAttemptSettings;
 import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.Futures;
 import io.grpc.CallOptions;
-import io.grpc.Status;
-import java.util.concurrent.CancellationException;
-import java.util.concurrent.Future;
-import java.util.concurrent.ThreadLocalRandom;
+import io.grpc.Status.Code;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
-import javax.annotation.Nullable;
 import org.joda.time.Duration;
 
 /**
- * Implements the retry and timeout functionality used in {@link UnaryCallable}. The behavior is
- * controlled by the given {@link RetrySettings}.
+ * Implements the retry and timeout functionality used in {@link UnaryCallable}.
+ *
+ * <p>
+ * The behavior is controlled by the given {@link RetrySettings}.
  */
 class RetryingCallable<RequestT, ResponseT> implements FutureCallable<RequestT, ResponseT> {
-
   // Duration to sleep on if the error is DEADLINE_EXCEEDED.
   static final Duration DEADLINE_SLEEP_DURATION = Duration.millis(1);
 
   private final FutureCallable<RequestT, ResponseT> callable;
-  private final RetrySettings retryParams;
-  private final UnaryCallable.Scheduler executor;
-  private final NanoClock clock;
+  private final RetryingExecutor<ResponseT> scheduler;
 
   RetryingCallable(
       FutureCallable<RequestT, ResponseT> callable,
       RetrySettings retrySettings,
-      UnaryCallable.Scheduler executor,
-      NanoClock clock) {
+      ScheduledExecutorService scheduler,
+      ApiClock clock) {
     this.callable = Preconditions.checkNotNull(callable);
-    this.retryParams = Preconditions.checkNotNull(retrySettings);
-    this.executor = executor;
-    this.clock = clock;
+    RetryAlgorithm retryAlgorithm =
+        new RetryAlgorithm(
+            new GrpcExceptionRetryAlgorithm(), new ExponentialRetryAlgorithm(retrySettings, clock));
+    this.scheduler = new ScheduledRetryingExecutor<>(retryAlgorithm, scheduler);
   }
 
   @Override
   public ApiFuture<ResponseT> futureCall(RequestT request, CallContext context) {
-    RetryingResultFuture resultFuture = new RetryingResultFuture();
-    context = getCallContextWithDeadlineAfter(context, retryParams.getTotalTimeout());
-    Retryer retryer =
-        new Retryer(
-            request,
-            context,
-            resultFuture,
-            retryParams.getInitialRetryDelay(),
-            retryParams.getInitialRpcTimeout(),
-            null);
-    resultFuture.issueCall(request, context, retryer);
-    return resultFuture;
+    GrpcRetryCallable<RequestT, ResponseT> retryCallable =
+        new GrpcRetryCallable<>(callable, request, context);
+
+    RetryingFuture<ResponseT> retryingFuture = scheduler.createFuture(retryCallable);
+    retryCallable.setExternalFuture(retryingFuture);
```

Comment:
You are right, it is circular and it is designed to be circular. This is forced by the current gax FutureCallables design. Please check my previous comment to this PR for more details what is happening here and why: https://github.com/googleapis/gax-java/pull/235#discussion_r106738328

(also the previous implementation of RetryingCallable was circular as well).